### PR TITLE
Extend autocompletion support

### PR DESCRIPTION
**Pull request by moverest**
Adds autocompletion support for: alsamixer, godoc, gofmt, mkdir, netctl-auto, poweroff, termite, udisksctl.

I made these for me but I figure there would be useful here. So there you have them if you want.

Edit: also: goimports, golint, gorename, xz, modinfo, lscpu.

### CODE REVIEW

File: share/completions/udisksctl.fish [hunk 1]
```diff
@@ -0,0 +1,67 @@
+set -l cmds help info dump status monitor mount unmount unlock lock loop-setup loop-delete power-off smart-simulate
+
+set __fish_mount_opts async\tUse\ asynchronous\ I/O atime\tUpdate\ time\ on\ each\ access auto\tMounted\ with\ -a defaults\tUse\ default\ options dev\tInterpret\ character/block\ special\ devices exec\tPermit\ executables _netdev\tFilesystem\ uses\ network noatime\tDo\ not\ update\ time\ on\ each\ access noauto\tNot\ mounted\ by\ -a nodev\tDo\ not\ interpret\ character/block\ special\ devices noexec\tDo\ not\ permit\ executables nosuid\tIgnore\ suid\ bits nouser\tOnly\ root\ may\ mount remount\tRemount\ read-only\ filesystem ro\tMount\ read-only rw\tMount\ read-write suid\tAllow\ suid\ bits sync\tUse\ synchronous\ I/O dirsync\tUse\ synchronous\ directory\ operations user\tAny\ user\ may\ mount users\tAny\ user\ may\ mount\ and\ unmount
+
+function __fish_complete_blockdevice
```
Comment (faho) [inline share/completions/udisksctl.fish]: This function has been split out recently (7af9e1f5c5bfe9b295309094fb03ef27ee6ca1fd). No need to redefine it here.

File: share/completions/udisksctl.fish [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    for f in $cmd*
+        [ -b $f ]
+        and printf "%s\t%s\n" $f "Block device"
+        [ -d $f ]
+        and printf "%s\n" $f/
+    end
+end
+
+function __fish_print_mounted_blockdevice
+    cat /proc/mounts | grep "^/.*" | cut -d " " -f 1
```
Comment (faho) [inline share/completions/udisksctl.fish]: Useless use of cat. Also `string` would be nice here.

File: share/completions/netctl-auto.fish [hunk 3]
```diff
@@ -0,0 +1,33 @@
+set -l cmds list store restore stop-all start stop restart switch-to status enable disable reenable is-enabled edit
+
+function __fish_print_netctl-auto_profile
+    for line in (netctl-auto list)
+        set -l profile (string sub -s 3 $line)
```
Comment (faho) [inline share/completions/netctl-auto.fish]: Where does the "3" come from? Does it print whitespace before the actual profile? If so, `string trim` might be nicer.

File: share/completions/netctl-auto.fish [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        if [ (string sub -l 1 $line) = "*" ]
```
Comment (faho) [inline share/completions/netctl-auto.fish]: `if string match -q '\**' -- $line`

File: share/completions/netctl-auto.fish (hunk 3 above)
Comment (moverest) [inline share/completions/netctl-auto.fish]: It prints either "*  " or "   ".

File: share/completions/udisksctl.fish (hunk 2 above)
Comment (moverest) [inline share/completions/udisksctl.fish]:
```
grep "^/.*"  /proc/mounts | cut -d " " -f 1`
```

I must admit that I am new to `string`. Should I do something like this?
```
string match -i -r "^/[^ ]*" < /proc/mounts
```

Comment (faho) [inline share/completions/udisksctl.fish]: Basically, yes. Though the "-i" is superfluous and you should `test -r /proc/mounts` before - you don't want errors here (this would also apply to `cat | grep`).

Of course the assumption here is that block devices can't have spaces in their names, which I _believe_ holds, and even if it didn't it would be a very unusual configuration. /etc/mtab replaces spaces with '\040', dunno if it's the same here.

File: share/completions/netctl-auto.fish (hunk 3 above)
Comment (faho) [inline share/completions/netctl-auto.fish]: In that case sub is okay.

File: share/completions/mkdir.fish [hunk 5]
```diff
@@ -0,0 +1,9 @@
+complete -c mkdir -s m -l mode --description 'Set file mode (as in chmod)' -x
+complete -c mkdir -s p -l parents --description 'Make parent directories as needed'
+complete -c mkdir -s v -l verbose --description 'Print a message for each created directory'
+complete -c mkdir -s Z --description 'Set SELinux security context of each created directory to the default type'
```
Comment (floam) [inline share/completions/mkdir.fish]: On OS X or BSD a few of these on a very prominent command like mkdir might be confusing to a user.

```
     mkdir [-pv] [-m mode] directory_name ...
```

The full manpage:

```
SYNOPSIS
     mkdir [-pv] [-m mode] directory_name ...

DESCRIPTION
     The mkdir utility creates the directories named as operands, in the order
     specified, using mode rwxrwxrwx (0777) as modified by the current
     umask(2).

     The options are as follows:

     -m mode
             Set the file permission bits of the final created directory to
             the specified mode.  The mode argument can be in any of the for-
             mats specified to the chmod(1) command.  If a symbolic mode is
             specified, the operation characters ``+'' and ``-'' are inter-
             preted relative to an initial mode of ``a=rwx''.

     -p      Create intermediate directories as required.  If this option is
             not specified, the full path prefix of each operand must already
             exist.  On the other hand, with this option specified, no error
             will be reported if a directory given as an operand already
             exists.  Intermediate directories are created with permission
             bits of rwxrwxrwx (0777) as modified by the current umask, plus
             write and search permission for the owner.

     -v      Be verbose when creating directories, listing them as they are
             created.

     The user must have write permission in the parent directory.

DIAGNOSTICS
     The mkdir utility exits 0 on success, and >0 if an error occurs.

SEE ALSO
     rmdir(1)

COMPATIBILITY
     The -v option is non-standard and its use in scripts is not recommended.

STANDARDS
     The mkdir utility is expected to be IEEE Std 1003.2 (``POSIX.2'') compat-
     ible.
```

Comment (moverest) [inline share/completions/mkdir.fish]: In this case, I can do something like this:

```fish
if test (uname) = "Linux"
    complete -c mkdir -s Z --description 'Set SELinux security context of each created directory to the default type'
    complete -c mkdir -l context --description 'Like -Z' -f
end
```

Or I can get rid of the file.

Comment (faho) [inline share/completions/mkdir.fish]: What has worked well for us is something like

```fish
if mkdir --version >/dev/null ^/dev/null
# GNU
else
# BSD/OSX
```

this will also work if you install GNU tools on OSX, and will also work on Hurd.

I think the only place we actually check `uname` for completions is killall, but that's a special case (you can't _ever_ invoke it on Solaris).

Comment (moverest) [inline share/completions/mkdir.fish]: Only `killall`, `modprobe` and `open` indeed. I'll use `mkdir --version` then.

I guess it's okay if I use `uname` for `modinfo` completions? I have a few other command completions done.

Comment (floam) [inline share/completions/mkdir.fish]: Yeah, testing for a feature directly is usually the way to go. 

I think perhaps further you should also test, somehow, for SELinux on Linux, and if that doesn't exist, don't show the --context=CTX (-Z CTX) options.

Comment (floam) [inline share/completions/mkdir.fish]: > I guess it's okay if I use uname for modinfo completions? I have a few other command completions done.

If there is no other OS that has a modinfo, seems like the thing to do.

Comment (floam) [inline share/completions/mkdir.fish]: Seems solaris does:

http://docs.oracle.com/cd/E23824_01/html/821-1462/modinfo-1m.html

But if you can only manage to  provide completions relevant to a Linux modinfo, it seems reasonable to check uname.

Comment (faho) [inline share/completions/mkdir.fish]: >But if you can only manage to provide completions relevant to a Linux modinfo, it seems reasonable to check uname. 

It seems like the linux one has `--version` and the solaris one doesn't - long options started as a GNUism, so other OSs don't have them, and if they do they're most likely compatible with the GNU stuff.

We should really try not to call uname - as I said, hurd will have GNU tools but won't identify as linux.

>I think perhaps further you should also test, somehow, for SELinux on Linux (or, at least, a SELinux-enabled mkdir), and if that doesn't exist, don't show the --context=CTX (-Z CTX) options. 

It seems like the canonical way to check if SELinux is enabled is `sestatus`. Since I think `mkdir` can also add contexts even if selinux is installed but not enabled, we should get away with just checking `command -s sestatus`.

Comment (floam) [inline share/completions/mkdir.fish]: > It seems like the linux one has --version and the solaris one doesn't - long options started as a GNUism, so other OSs don't have them, and if they do they're most likely compatible with the GNU stuff.

They seem to be more different than that?

http://docs.oracle.com/cd/E23824_01/html/821-1462/modinfo-1m.html

Comment (moverest) [inline share/completions/mkdir.fish]: Seems like `modinfo` is behave in a completely different way on Solaris... :-/

If I understand well, I should have for `mkdir`:

```fish
if command -s sestatus
    complete -c mkdir -s Z --description 'Set SELinux security context of each created directory to the default type'
    complete -c mkdir -l context --description 'Like -Z' -f
end
```

But, `--version` is only present with GNU tools, so I should also have:
```fish
if mkdir --version >/dev/null ^/dev/null
    complete -c mkdir -l version --description 'Output version'
end
```

Comment (faho) [inline share/completions/mkdir.fish]: >They seem to be more different than that?

Sure, but that doesn't matter here - if we have `--version`, then we have the GNU version. If we don't, then it's something else (currently it only seems to be a thing on Solaris).

>If I understand well, I should have for mkdir:

Yes, but silence `command`s stdout - `>/dev/null`.

>But, --version is only present with GNU tools, so I should also have:

Nope - the entire completion here is specific to the GNU version. The BSD version just has `-m`, `-p` and `-v` - just those short options, not the long versions.

So it's more like

```fish
if mkdir --version >/dev/null ^/dev/null
    complete -c mkdir -l version -d 'Output version'
    complete -c mkdir -s m -l mode --description 'Set file mode (as in chmod)' -x
    # [...]
else
    complete -c mkdir -s m --description 'Set file mode (as in chmod)' -x
    # [...]
end
```

In other words: Use how it reacts to "--version" as an indicator.

File: share/completions/modinfo.fish [hunk 6]
```diff
@@ -0,0 +1,16 @@
+function __fish_print_modules
```
Comment (faho) [inline share/completions/modinfo.fish]: This is the same thing as in the modprobe completions. It should be split out into its own file in share/functions.

File: share/completions/mkdir.fish [hunk 7]
```diff
@@ -1,9 +1,20 @@
...
+else
+    complete -c mkdir -s m --description 'Set file mode (as in chmod)' -x
+    complete -c mkdir -s p --description 'Make parent directories as needed'
+    complete -c mkdir -s v --description 'Print a message for each created directory'
+end
+
+# Checks if SELinux is installed
+if command -s sestatus > /dev/null ^ /dev/null
+    complete -c mkdir -s Z --description 'Set SELinux security context of each created directory to the default type'
+    complete -c mkdir -l context --description 'Like -Z' -f
```
Comment (floam) [inline share/completions/mkdir.fish]: I see that  -Z is the short option for --context -- so instead of giving it a description of "Like -Z", just let it have the same description and have fish show the options together .

**Comment by faho** [comment]
I've merged this now. Since there's a bunch of commits, I feel a merge commit is appropriate here.

Thanks!
